src/main/java/de/srendi/advancedperipherals/common/util/inventory/ChemicalUtil.java

```diff
@@ -0,0 +1,132 @@
+            return itemHandler;
+        if (object instanceof BlockEntity blockEntity && level == null && pos == null) {
+            pos = blockEntity.getBlockPos();
+            level = blockEntity.getLevel();
+        }
+        if (level != null && pos != null) {
+            return level.getCapability(Capabilities.CHEMICAL.block(), pos, direction != null ? direction : Direction.NORTH);
+        }
+        return null;
+    }
+
+    @Nullable
+    public static IChemicalHandler getHandlerFromDirection(@NotNull String direction, @NotNull IPeripheralOwner owner) throws LuaException {
+        Level level = owner.getLevel();
+        Objects.requireNonNull(level);
+        Direction relativeDirection = CoordUtil.getDirection(owner.getOrientation(), direction);
+        BlockEntity target = level.getBlockEntity(owner.getPos().relative(relativeDirection));
+        if (target == null)
+            return null;
+
+        return extractHandler(target, level, owner.getPos().relative(relativeDirection), relativeDirection);
+    }
+
+    @Nullable
+    public static IChemicalHandler getHandlerFromName(@NotNull IComputerAccess access, String name) throws LuaException {
+        IPeripheral location = access.getAvailablePeripheral(name);
+
+        // Tanks/Block Entities can't be accessed if the bridge is not exposed to the same network as the target tank/block entity
+        // This can occur when the bridge was wrapped via a side and not via modems
+        if (location == null)
+            return null;
+
+        IChemicalHandler handler = extractHandler(location.getTarget(), null, null, null);
+        if (handler == null)
+            throw new LuaException("Target '" + name + "' is not a fluid handler");
+        return handler;
+    }
+
+    public static ChemicalStack toChemicalStack(Chemical chemical, long amount) {
+        return new ChemicalStack(MekanismAPI.CHEMICAL_REGISTRY.wrapAsHolder(chemical), amount);
+    }
+
+    @NotNull
+    public static String getFingerprint(@NotNull ChemicalStack stack) {
+        // A pretty lame fingerprint, a chemical stack does not have any components or other stuff
+        String fingerprint = getRegistryKey(stack).toString();
+        try {
+            byte[] bytesOfHash = fingerprint.getBytes(StandardCharsets.UTF_8);
+            MessageDigest md = MessageDigest.getInstance("MD5");
+            return StringUtil.toHexString(md.digest(bytesOfHash));
+        } catch (NoSuchAlgorithmException ex) {
+            AdvancedPeripherals.debug("Could not parse fingerprint.", org.apache.logging.log4j.Level.ERROR);
+            ex.printStackTrace();
+        }
+        return "";
+    }
+
+    public static ResourceLocation getRegistryKey(Chemical fluid) {
+        return MekanismAPI.CHEMICAL_REGISTRY.getKey(fluid);
+    }
+
+    public static ResourceLocation getRegistryKey(ChemicalStack fluid) {
+        return MekanismAPI.CHEMICAL_REGISTRY.getKey(fluid.copy().getChemical());
```

Comment:
do we need copy to get chemical here?

Reply:
No, I copied it from mekanism and I didn't noticed it